src/snakeoil/compression/_xz.py

```diff
@@ -0,0 +1,72 @@
+"""
+xz decompression/compression
+
+Where possible, this module defers to cpython's lzma module - if it's not available,
+it defers to executing xz with tempfile arguments to do decompression
+and compression.
+
+Use this module unless it's absolutely critical that lzma module be used.
+"""
+
+__all__ = ("compress_data", "decompress_data")
+
+import multiprocessing
+from functools import partial
+
+from .. import process
+from ..compression import _util
+
+# Unused import
+# pylint: disable=W0611
+
+# if xz can't be found, throw an error.
+xz_path = process.find_binary("xz")
+xz_compress_args = ('-T%i' % multiprocessing.cpu_count(),)
+xz_decompress_args = xz_compress_args
+parallelizable = True
+
+try:
+    from lzma import LZMAFile
+    from lzma import compress as _compress_data
+    from lzma import decompress as _decompress_data
+    native = True
+except ImportError:
+
+    # We need this because if we are not native then TarFile.open will fail
+    # (and some code needs to be able to check that).
+    native = False
+
+    _compress_data = partial(_util.compress_data, xz_path)
+    _decompress_data = partial(_util.decompress_data, xz_path)
+
+_compress_handle = partial(_util.compress_handle, xz_path)
+_decompress_handle = partial(_util.decompress_handle, xz_path)
+
+
+def compress_data(data, level=9, parallelize=False):
+    if parallelize and parallelizable:
+        return _util.compress_data(xz_path, data, compresslevel=level,
+                                   extra_args=xz_compress_args)
+    return _compress_data(data, compresslevel=level)
+
+def decompress_data(data, parallelize=False):
+    if parallelize and parallelizable:
+        return _util.decompress_data(xz_path, data,
+                                     extra_args=xz_decompress_args)
+    return _decompress_data(data)
+
+def compress_handle(handle, level=9, parallelize=False):
+    if parallelize and parallelizable:
+        return _util.compress_handle(xz_path, handle, compresslevel=level,
+                                     extra_args=xz_compress_args)
+    elif native and isinstance(handle, str):
+        return LZMAFile(handle, mode='w', compresslevel=level)
```

Comment:
Are you sure it supports `compresslevel` argument? Checking the implementation of `LZMAFile`, I don't see any mention of `compresslevel`

Reply:
> I see that there are no tests at all for compression part. The closest thing is the tests for `bz2_source`, but I believe it would be useless to add equivalent `xz_source`.
> 
> For now, it would be enough if we have manual testing of code, in those configurations:
> 
>     1. Using native python `LZMAFile`
> 
>     2. Using `xz -T...` with old version of `xz` (to check that it works and ignores the "unknown" arg)
> 
>     3. Using `xz -T...` with latest version of `xz` (to check the feature works)

I need to address your comments, but I've tested both 2 & 3 which are fine, but I think I need a proper way to invoke this module manually for testing. :(

Reply:
I didn't mean to make it hard for you. I **don't** expect you to create unit tests!

I wanted to know if it works for you when you force the native stdlib variant (for example by changing the `if`s in code).